File: internal/agent/protocol.go
package agent

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
)

type MessageType string

const (
	MessageTypeHello         MessageType = "HELLO"
	MessageTypePause         MessageType = "PAUSE"
	MessageTypeResume        MessageType = "RESUME"
	MessageTypeState         MessageType = "STATE"
	MessageTypeExtractResult MessageType = "EXTRACT_RESULT"
	MessageTypeGoodbye       MessageType = "GOODBYE"
)

const MaxFrameSize = 16 * 1024 * 1024

type Message struct {
	Type MessageType `json:"type"`
	ID   uint64      `json:"id,omitempty"`

	Version    string `json:"version,omitempty"`
	Platform   string `json:"platform,omitempty"`
	AppPackage string `json:"app_package,omitempty"`

	Snapshots map[string]json.RawMessage `json:"snapshots,omitempty"`

	Extractor string          `json:"extractor,omitempty"`
	Result    json.RawMessage `json:"result,omitempty"`
	Error     string          `json:"error,omitempty"`

	Reason string `json:"reason,omitempty"`
}

func Hello(version, platform, appPackage string) Message {
	return Message{Type: MessageTypeHello, Version: version, Platform: platform, AppPackage: appPackage}
}

func Pause(id uint64) Message { return Message{Type: MessageTypePause, ID: id} }

func Resume(id uint64) Message { return Message{Type: MessageTypeResume, ID: id} }

func State(id uint64, snapshots map[string]json.RawMessage) Message {
	return Message{Type: MessageTypeState, ID: id, Snapshots: snapshots}
}

func ExtractResult(id uint64, extractor string, result json.RawMessage, extractorError string) Message {
	return Message{
		Type:      MessageTypeExtractResult,
		ID:        id,
		Extractor: extractor,
		Result:    result,
		Error:     extractorError,
	}
}

func Goodbye(reason string) Message {
	return Message{Type: MessageTypeGoodbye, Reason: reason}
}

func WriteMessage(writer io.Writer, message Message) error {
	payload, err := json.Marshal(message)
	if err != nil {
		return fmt.Errorf("marshal: %w", err)
	}
	if len(payload) > MaxFrameSize {
		return fmt.Errorf("frame of %d bytes exceeds maximum %d", len(payload), MaxFrameSize)
	}
	var header [4]byte
	binary.BigEndian.PutUint32(header[:], uint32(len(payload)))
	if _, err := writer.Write(header[:]); err != nil {
		return fmt.Errorf("write header: %w", err)
	}
	if _, err := writer.Write(payload); err != nil {
		return fmt.Errorf("write payload: %w", err)
	}
	return nil
}

func ReadMessage(reader io.Reader) (Message, error) {
	var header [4]byte
	if _, err := io.ReadFull(reader, header[:]); err != nil {
		return Message{}, err
	}
	length := binary.BigEndian.Uint32(header[:])
	if length > MaxFrameSize {
		return Message{}, fmt.Errorf("frame of %d bytes exceeds maximum %d", length, MaxFrameSize)
	}
	payload := make([]byte, length)
	if _, err := io.ReadFull(reader, payload); err != nil {
		return Message{}, fmt.Errorf("read payload: %w", err)
	}
	var message Message
	if err := json.Unmarshal(payload, &message); err != nil {
		return Message{}, fmt.Errorf("unmarshal: %w", err)
	}
	if message.Type == "" {
		return Message{}, errors.New("missing type")
	}
	return message, nil
}

File: sdk/android/src/main/kotlin/dev/uatu/sdk/Protocol.kt
package dev.uatu.sdk

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import org.json.JSONArray
import org.json.JSONObject

enum class MessageType(val wire: String) {
    HELLO("HELLO"),
    PAUSE("PAUSE"),
    RESUME("RESUME"),
    STATE("STATE"),
    EXTRACT_RESULT("EXTRACT_RESULT"),
    GOODBYE("GOODBYE");

    companion object {
        fun fromWire(wire: String): MessageType =
            values().firstOrNull { it.wire == wire }
                ?: throw IOException("unknown message type: $wire")
    }
}

data class Message(
    val type: MessageType,
    val id: Long = 0L,
    val version: String? = null,
    val platform: String? = null,
    val appPackage: String? = null,
    val snapshots: Map<String, Any?>? = null,
    val extractor: String? = null,
    val result: Any? = null,
    val error: String? = null,
    val reason: String? = null,
) {
    companion object {
        fun hello(version: String, platform: String, appPackage: String): Message =
            Message(MessageType.HELLO, version = version, platform = platform, appPackage = appPackage)

        fun pause(id: Long): Message = Message(MessageType.PAUSE, id = id)

        fun resume(id: Long): Message = Message(MessageType.RESUME, id = id)

        fun state(id: Long, snapshots: Map<String, Any?>): Message =
            Message(MessageType.STATE, id = id, snapshots = snapshots)

        fun extractResult(id: Long, extractor: String, result: Any?, error: String? = null): Message =
            Message(MessageType.EXTRACT_RESULT, id = id, extractor = extractor, result = result, error = error)

        fun goodbye(reason: String): Message = Message(MessageType.GOODBYE, reason = reason)
    }
}

object Protocol {
    const val MAX_FRAME_SIZE: Int = 16 * 1024 * 1024

    @Throws(IOException::class)
    fun write(output: OutputStream, message: Message) {
        val bytes = toJson(message).toString().toByteArray(Charsets.UTF_8)
        if (bytes.size > MAX_FRAME_SIZE) {
            throw IOException("frame of ${bytes.size} bytes exceeds maximum $MAX_FRAME_SIZE")
        }
        DataOutputStream(output).apply {
            writeInt(bytes.size)
            write(bytes)
            flush()
        }
    }

    @Throws(IOException::class)
    fun read(input: InputStream): Message {
        val dataInput = DataInputStream(input)
        val length = dataInput.readInt()
        if (length < 0 || length > MAX_FRAME_SIZE) {
            throw IOException("frame of $length bytes exceeds maximum $MAX_FRAME_SIZE")
        }
        val bytes = ByteArray(length)
        dataInput.readFully(bytes)
        return fromJson(JSONObject(String(bytes, Charsets.UTF_8)))
    }

    private fun toJson(message: Message): JSONObject {
        val json = JSONObject()
        json.put("type", message.type.wire)
        if (message.id != 0L) json.put("id", message.id)
        message.version?.let { json.put("version", it) }
        message.platform?.let { json.put("platform", it) }
        message.appPackage?.let { json.put("app_package", it) }
        message.snapshots?.let { snapshots ->
            val snapshotsJson = JSONObject()
            for ((key, value) in snapshots) {
                snapshotsJson.put(key, wrap(value))
            }
            json.put("snapshots", snapshotsJson)
        }
        message.extractor?.let { json.put("extractor", it) }
        message.result?.let { json.put("result", wrap(it)) }
        message.error?.let { json.put("error", it) }
        message.reason?.let { json.put("reason", it) }
        return json
    }

    private fun fromJson(json: JSONObject): Message {
        val typeString = json.optString("type", "")
        if (typeString.isEmpty()) throw IOException("missing type")
        return Message(
            type = MessageType.fromWire(typeString),
            id = json.optLong("id", 0L),
            version = json.optStringOrNull("version"),
            platform = json.optStringOrNull("platform"),
            appPackage = json.optStringOrNull("app_package"),
            snapshots = json.optJSONObject("snapshots")?.let { snapshotsJson ->
                snapshotsJson.keys().asSequence().associateWith { unwrap(snapshotsJson.get(it)) }
            },
            extractor = json.optStringOrNull("extractor"),
            result = if (json.has("result") && !json.isNull("result")) unwrap(json.get("result")) else null,
            error = json.optStringOrNull("error"),
            reason = json.optStringOrNull("reason"),
        )
    }

    private fun wrap(value: Any?): Any = when (value) {
        null -> JSONObject.NULL
        is Number, is Boolean, is String -> value
        is Map<*, *> -> JSONObject().also { json ->
            for ((key, nested) in value) json.put(key.toString(), wrap(nested))
        }
        is List<*> -> JSONArray().also { array ->
            for (item in value) array.put(wrap(item))
        }
        else -> value.toString()
    }

    private fun unwrap(value: Any?): Any? = when (value) {
        JSONObject.NULL, null -> null
        is JSONObject -> value.keys().asSequence().associateWith { unwrap(value.get(it)) }
        is JSONArray -> buildList { for (index in 0 until value.length()) add(unwrap(value.get(index))) }
        else -> value
    }

    private fun JSONObject.optStringOrNull(key: String): String? =
        if (has(key) && !isNull(key)) getString(key) else null
}
